lib/MadNLPGPU/src/cusolver.jl

```diff
@@ -0,0 +1,180 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+const CuSubVector{T} = SubArray{T, 1, CUDA.CuArray{T, 1, CUDA.Mem.DeviceBuffer}, Tuple{CUDA.CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}, false}
+
+#=
+    cusolverRF
+=#
+
+@kwdef mutable struct RFSolverOptions <: MadNLP.AbstractOptions
+    rf_symbolic_analysis::Symbol = :klu
+    rf_fast_mode::Bool = true
+    rf_pivot_tol::Float64 = 1e-14
+    rf_boost::Float64 = 1e-14
+    rf_factorization_algo::CUSOLVER.cusolverRfFactorization_t = CUSOLVER.CUSOLVERRF_FACTORIZATION_ALG0
+    rf_triangular_solve_algo::CUSOLVER.cusolverRfTriangularSolve_t = CUSOLVER.CUSOLVERRF_TRIANGULAR_SOLVE_ALG2
+end
+
+mutable struct RFSolver{T} <: MadNLP.AbstractLinearSolver{T}
+    inner::Union{Nothing,CUSOLVERRF.RFLowLevel}
+
+    tril::CUSPARSE.CuSparseMatrixCSC{T}
+    full::CUSPARSE.CuSparseMatrixCSR{T}
+    tril_to_full_view::CuSubVector{T}
+    dummy::CUDA.CuVector{T}
+
+    opt::RFSolverOptions
+    logger::MadNLP.MadNLPLogger
+end
+
+function RFSolver(
+    csc::CUSPARSE.CuSparseMatrixCSC;
+    opt=RFSolverOptions(),
+    logger=MadNLP.MadNLPLogger(),
+)
+    n, m = size(csc)
+    @assert n == m
+
+    full,tril_to_full_view = MadNLP.get_tril_to_full(csc)
+
+    full = CUSPARSE.CuSparseMatrixCSR(
+        full.colPtr,
+        full.rowVal,
+        full.nzVal,
+        full.dims
+    )
+
+    return RFSolver(
+        nothing, csc, full, tril_to_full_view, similar(csc.nzVal,1),
+        opt, logger
+    )
+end
+
+function MadNLP.factorize!(M::RFSolver)
+    copyto!(M.full.nzVal, M.tril_to_full_view)
+    if M.inner == nothing
+        sym_lu = CUSOLVERRF.klu_symbolic_analysis(M.full)
+        M.inner = CUSOLVERRF.RFLowLevel(
+            sym_lu;
+            fast_mode=M.opt.rf_fast_mode,
+            factorization_algo=M.opt.rf_factorization_algo,
+            triangular_algo=M.opt.rf_triangular_solve_algo,
+            nboost=M.opt.rf_boost,
+            nzero=M.opt.rf_pivot_tol,
+        )
+    end
+    CUSOLVERRF.rf_refactor!(M.inner, M.full)
+    return M
+end
+
+function MadNLP.solve!(M::RFSolver{T}, x) where T
+    CUSOLVERRF.rf_solve!(M.inner, x)
+    # this is necessary to not distort the timing in MadNLP
+    copyto!(M.dummy, M.dummy)
+    synchronize(CUDABackend())
```

Comment:
Yeah, it needed some operation on GPU and then synchronize. Maybe `CUDA.sumchronize()` could have been different? Let me check.

Reply:
It seems that we need to run something on GPU until we can synchronize things. I'm suspecting that this is a CUDA's own issue; otherwise running `synchronize` should wait until the kernels are all run